### ISSUE TITLE
Remove Pillow dependency

### ISSUE THREAD

**Issue by cortesi** [mitmproxy/mitmproxy#1900]
Pillow is a large dependency with a platform-specific binary components. We only use it to extract EXIF data for content views, and there are a number of pure-Python modules that can do this for us instead. This looks like a fairly well-tested alternative:

https://github.com/hMatoba/Piexif

**Comment by lymanZerga11** [comment]
Don't you think [exifread](https://github.com/ianare/exif-py) is a better alternative than Piexif?
Piexif gives a lot of the information in integer codes which should be decoded against the standards mentioned in [this document](http://www.cipa.jp/std/documents/e/DC-008-2012_E.pdf). 
Exifread gives the same information in a more human-readable encoding.

If you agree, I can start working on this issue.
Here's a comparison between sample data from Piexif and exifread :

```
<< Test of exifread >>

EXIF BodySerialNumber = H52502123
EXIF CVAPattern = [2, 0, 2, 0, 0, 1, 1, 2]
EXIF ColorSpace = Uncalibrated
EXIF Contrast = Normal
EXIF CustomRendered = Normal
EXIF DateTimeDigitized = 2011:06:09 01:20:59
EXIF DateTimeOriginal = 2011:06:09 01:20:59
EXIF DigitalZoomRatio = 1
EXIF ExifImageLength = 600
EXIF ExifImageWidth = 600
EXIF ExifVersion = 0230
EXIF ExposureBiasValue = -3/10
EXIF ExposureMode = Manual Exposure
EXIF ExposureProgram = Aperture Priority
EXIF ExposureTime = 1/1000
EXIF FileSource = Digital Camera
EXIF Flash = Flash did not fire, auto mode
EXIF FocalLength = 77
EXIF FocalLengthIn35mmFilm = 116
EXIF GainControl = Low gain up
EXIF ISOSpeedRatings = 200
EXIF LensMake = Pentax
EXIF LensModel = smc Pentax F A77 Limited
EXIF LightSource = Unknown
EXIF MaxApertureValue = 0
EXIF MeteringMode = Pattern
EXIF Saturation = Normal
EXIF SceneCaptureType = Standard
EXIF Sharpness = Normal
EXIF ShutterSpeedValue = 1245723/125000
EXIF UserComment = MY TEST COMMENT!
EXIF WhiteBalance = Auto
Image Artist = Robin Parmar
Image Copyright = Robin Parmar
Image DateTime = 2011:08:22 21:39:05
Image ExifOffset = 11444
Image ImageDescription =
Image Make = OLYMPUS IMAGING CORP.
Image Model = E-P1
Image Orientation = Horizontal (normal)
Image ResolutionUnit = Pixels/Inch
Image Software = Adobe Photoshop CS5 Windows
Image XResolution = 72
Image YResolution = 72
Thumbnail Compression = JPEG (old-style)
Thumbnail JPEGInterchangeFormat = 878
Thumbnail JPEGInterchangeFormatLength = 10416
Thumbnail ResolutionUnit = Pixels/Inch
Thumbnail XResolution = 72
Thumbnail YResolution = 72
```

```
<< Test of Piexif >>

Exif:
36864 = b'0230'
37377 = (9965784, 1000000)
40962 = 600
36867 = b'2011:06:09 01:20:59'
36868 = b'2011:06:09 01:20:59'
37381 = (0, 256)
37510 = b'ASCII\x00\x00\x00MY TEST COMMENT!'
37383 = 5
37384 = 0
37385 = 24
37386 = (77, 1)
41988 = (100, 100)
41986 = 1
40963 = 600
37380 = (-3, 10)
41730 = b'\x02\x00\x02\x00\x00\x01\x01\x02'
33434 = (1, 1000)
41728 = b'\x03'
41989 = 116
34850 = 3
42033 = b'H52502123'
40961 = 65535
41990 = 0
34855 = 200
41987 = 0
41991 = 1
41992 = 0
42035 = b'Pentax'
42036 = b'smc Pentax F A77 Limited'
41993 = 0
41994 = 0
41985 = 0

0th:
283 = (720000, 10000)
296 = 2
34665 = 11444
306 = b'2011:08:22 21:39:05'
270 = b''
271 = b'OLYMPUS IMAGING CORP.'
272 = b'E-P1'
305 = b'Adobe Photoshop CS5 Windows'
274 = 1
33432 = b'Robin Parmar'
282 = (720000, 10000)
315 = b'Robin Parmar'

1st:
513 = 878
514 = 10416
259 = 6
296 = 2
282 = (72, 1)
283 = (72, 1)

GPS:

Interop:
```

**Comment by AP-whitehat** [comment]
What is the status of this issue . 
I would like to work on this issue .
what about the debate on Piexif and exifread ?

**Comment by cortesi** [comment]
Hi there - exifread looks like a fine, well-supported library, and I agree the data seems easier to work with.  Let's do it! :)

**Comment by mohanKrishna97** [comment]
I like to fix this issue . I started working on it now.

**Comment by AP-whitehat** [comment]
First time here ..
@cortesi (or someone) could you tell me which portions of mitmproxy uses Pillow ?

**Comment by lymanZerga11** [comment]
Hi @AP-whitehat - Pillow is a better maintained fork of the Python Imaging Library (PIL). To maintain backwards compatibility, `PIL` module name is used in imports. As @cortesi has mentioned, it is mainly used for the extraction of EXIF data for content views (hence `exifread`). So most of the modifications will be in [/mitmproxy/mitmproxy/contentviews/image.py](https://github.com/mitmproxy/mitmproxy/blob/8112bce42413bfb056022002667ccda4090df615/mitmproxy/contentviews/image.py)
To completely remove dependency, however, you might have to modify some of the tests and examples as well. 
I think @mohanKrishna97 has started working on it. Any updates?

**Comment by Kriechi** [comment]
Seems like `exifread` only works with TIFF and JPG files. I could not get it to parse any data out of GIF, PNG, or ICO images.

**Comment by Kriechi** [comment]
Same is true for piexif.
Any more suggestions?

**Comment by mhils** [comment]
Persisting what I just said on Slack: I'm not fundamentally opposed against Pillow, but I see three downsides:

- Pillow is big
- Pillow has major native dependencies.
- I dislike passing random bytes to C code.

The lack of GIF/PNG/ICO support is annoying. For PNG and GIF it should be very easy to write a parser that extracts the metadata (GSoC?), I have no idea about ICO or WEBP. Given that this still is a bit of work, I'm personally undecided between keeping Pillow or reinventing the wheel in Python - it's clearly not as one-sided as in the lxml case.

**Comment by s4chin** [comment]
I am working on this.

**Comment by dequis** [comment]
This was linked from the gsoc ideas list, but it has already been fixed by #2028

**Comment by mhils** [comment]
Thanks, I updated the ideas page. :)
